libraries/botbuilder-core/src/privateConversationState.ts

```
@@ -0,0 +1,71 @@
+ * providers.
+ *
+ * ```JavaScript
+ * const { PrivateConversationState, MemoryStorage } = require('botbuilder');
+ *
+ * const PrivateConversationState = new PrivateConversationState(new MemoryStorage());
+ * adapter.use(PrivateConversationState);
+ *
+ * server.post('/api/messages', (req, res) => {
+ *    adapter.processActivity(req, res, async (context) => {
+ *       // Get loaded PrivateConversation state
+ *       const convo = PrivateConversationState.get(context);
+ *
+ *       // ... route activity ...
+ *
+ *    });
+ * });
+ * ```
+ */
+export class PrivateConversationState extends BotState {
+    /**
+     * Creates a new PrivateConversationState instance.
+     * @param storage Storage provider to persist PrivateConversation state to.
+     * @param namespace (Optional) namespace to append to storage keys. Defaults to an empty string.
+     */
+    constructor(storage: Storage, private namespace: string = '') {
+        super(storage, (context: TurnContext) => {
+            // Calculate storage key
+            const key: string = this.getStorageKey(context);
+            return key ? Promise.resolve(key) : Promise.reject(new Error(NO_KEY));
+        });
+    }
+
+    /**
+     * Returns the storage key for the current PrivateConversation state.
+     * @param context Context for current turn of PrivateConversation with the user.
+     */
+    public getStorageKey(context: TurnContext): string | undefined {
+        const activity: Activity = context.activity;
+        const channelId: string = activity.channelId;
+        const conversationId: string = activity && activity.conversation && activity.conversation.id ? activity.conversation.id : undefined;
+        const userId: string = activity && activity.from && activity.from.id ? activity.from.id : undefined;
+
+        return channelId && conversationId && userId ? `${channelId}/conversations/${conversationId}/users/${userId}/${this.namespace}` : undefined;
```

Its scary that this returns undefined. That seems like it would corrupt a data store, as it would just start writing things to the null key. 

Consider throwing an exception rather than returning undefined.